django/core/files/storage/base.py

```diff
@@ -51,6 +51,9 @@ def save(self, name, content, max_length=None):
         validate_file_name(name, allow_relative_path=True)
         return name
 
+    def is_name_available(self, name, max_length=None):
+        return not self.exists(name) and not (max_length and len(name) > max_length)
```

Comment:
Perhaps it's a matter of personal taste, and while the proposed code certainly follows the previous code, I think the following is slightly more correct because otherwise for falsey `max_length` values (such as 0)  we are not enforcing any `length`:
```suggestion
        return not self.exists(name) and (max_length is None or len(name) <= max_length)
```

Reply:
On a second thought, perhaps we need to leave this as is and create a new ticket to decide whether a `max_length` of `0` means no validation or what does it mean. Specifically, in `main` (and in this branch), this test pass:

```python
    def test_file_name_truncation_max_length_zero(self):
        name = "test_long_file_name.txt"
        file = ContentFile(b"content")
        stored_name = self.storage.save(name, file, max_length=0)
        self.assertEqual(stored_name, "test_long_file_name.txt")
```

I would have expected for it to fail somehow :thinking:

Reply:
I think max_length mostly comes from max_length of FileField which does have validation that it must be a positive integer (check code `fields.E121`)

Reply:
Yes, absolutely, I agree that is how it was incorporated (having look at the git history). I still think the `Storages` API should also check or do something with a `max_length` with value `0` since it's a public API "on its own", independently of Django fields. If you agree I'll create the ticket, or perhaps better to post to the forum first? :thinking: 
In any case, nothing within the scope of this PR, which looks amazing. Thank you!